docs/developer-guide/ksqldb-clients/contributing.md

```diff
@@ -0,0 +1,112 @@
+- Do not block. Rather, perform an action asynchronously on each new row as it arrives
+- Block until the next row arrives
+- Block until all rows have arrived. This only makes sense in the context of pull queries and push queries with `LIMIT` clauses. 
+
+Details of client interfaces to support these methods will likely vary by language. For example, the [Java client](java-client.md)
+uses [Reactive Streams](http://www.reactive-streams.org/) for its asynchronous streaming interface. 
+
+### Terminating push queries ###
+
+Requests to terminate push queries should be sent to the [`/close-query` server endpoint](TODO).
+The client should expose an interface that allows users to pass in a push query ID to be passed to this endpoint.
+
+### Inserting new rows of data into existing ksqlDB streams ###
+
+ksqlDB supports inserting new rows of data into existing ksqlDB streams via the [`/inserts-stream` server endpoint](TODO).
+If the server successfully validates the target stream name from a request made to the `/inserts-stream` endpoint,
+the client may then write JSON-serialized rows of data to the connection, and the server will insert each
+asynchronously and write an ack containing the relevant sequence number to the response.
+
+Because the insertions happen asynchronously on the server side, it makes sense for the client to expose APIs
+for inserting one row at a time, but inserting batches of rows is less useful due to the lack of transactional guarantees.
+
+The client should provide a convenient way for users to create rows of data to insert.
+The Java client's [KsqlObject](TODO) is one example, though details will likely vary by language.
+
+### Creation of new streams, tables, and persistent queries ###
+
+These DDL/DML requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
+
+### Terminating persistent queries ###
+
+These requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
+
+### Listing existing streams, tables, topics, and queries ###
+
+These admin operation requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
+
+Configurations
+--------------
+
+The client of course needs to expose options for specifying the address of the ksqlDB server to connect to.
+Additional options important to expose include: 
+- Support for TLS-enabled ksqlDB servers
+- Support for mutual-TLS-enabled ksqlDB servers
+- Support for ksqlDB servers with HTTP basic authentication enabled
+
+It may also be nice to support configuring multiple ksqlDB server addresses, so the client may route requests
+to a different server if one is down.
+
+Process
+-------
+
+Thanks for your interest in contributing a client!
+
+To get started:
+- Open [KLIP](../../../design-proposals/README.md) to propose the client you'd like to implement. The KLIP should include a high-level design and example interfaces.
```

Comment:
Is the expectation that if someone wants to tackle a client, they're expected to complete the project (implement all the functionality listed above)? Seems like quite the commitment. Might be good to add a sentence clarifying the expectation one way or the other so readers aren't left wondering.

Reply:
I would say a client can choose to implement some or all of the functionality.  This will become more important as new features are added to the client API.

This allows contributes to submit clients that just support a minimal subset of operations, rather than putting them off by saying 'it's all or nothing mate'. Such minimal clients can be extended by others as required.

Of course, we can say that anyone supporting push queries _must_ support terminating them as well.

Side question: if the client doesn't terminate the query, but just disconnects, are the queries terminated?

Reply:
> Side question: if the client doesn't terminate the query, but just disconnects, are the queries terminated?

Push queries are terminated when the underlying connection is closed, or by the server after 10 minutes of unresponsiveness from the client.